File: Algorithm1.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Created by s130973 on 17-3-2017.
 */
public class Algorithm1 {
    List<Pair> cpt = new ArrayList<>();

    public List sampleIndex(List<Tuple> S, Index I, int n) {
        for (Tuple t : S) {
            int count = lookup(t);

            Pair p = new Pair(t, count);
            cpt.add(p);
        }

        int sum = 0;
        for (Pair p : cpt){
            sum += p.getCount();
        }
        System.out.println(sum);

        List<Tuple> Sout = new ArrayList();

        List<Integer> sid = sampleRandomIntegers(n, sum);
        for (int id : sid) {
            int chosen = chosenTupleId(id, n);
            Tuple tS = cpt.get(chosen).getTuple();
            int offset = id - offsetId(chosen);
            Tuple tA = I.lookup(tS);
            Sout.add(tA);
            Sout.add(tS);
        }
        return Sout
    }

    public int lookup(Tuple t) {
        return 99;
    }

    //this method gives unique random integers with size Min(n, sum)
    private List<Integer> sampleRandomIntegers(int n, int sum) {
        int min = Math.min(sum, n);
        List<Integer> shuffleList = new ArrayList<>();
        for (int i = 0; i < sum; i++) {
            shuffleList.add(i,i);
        }
        Collections.shuffle(shuffleList);
        return shuffleList.subList(0, min);
    }

    private int chosenTupleId(int id, int n) {
        int sum = 0;
        for (int i = 0; i < n; i++) {
            sum += cpt.get(i).getCount();
            if (sum > id) {
                return i;
            }
        }
        System.out.println("Error in chosenTupleId");
        return -1;
    }
    private int offsetId(int chosen) {
        int sum = 0;
        if (chosen == 0) {
            return 0;
        }
        for (int i = 0; i < chosen; i++) {
            sum += cpt.get(i).getCount()
        }
        return sum;
    }
}

File: Algorithm1.py
import random

cpt = []


def sampleIndex(S, I, n):
    for t in S:
        count = lookup(t)
        print('t en count', t, count)

        p = (t, count)
        cpt.append(p)

    intSum = 0
    for p in cpt:
        #print('p[0]', p[0], 'p[1]', p[1])
        intSum += p[1]

    # print (intSum)

    Sout = []

    intMin = min(n, intSum)
    print('intMin', intMin, 'intSum', intSum)
    sid = random.sample(range(intSum), intMin)
    print('sid', sid)
    for id in sid:
        chosen = chosenTupleId(id, n)
        tS = cpt[chosen][0]
        # print('tS', tS)
        print('chosen', chosen)
        #intOff =
        #print('intOff', intOff)
        offset = id - offsetID(chosen)
        print('offset', offset)
        tA = lookupTuple(tS)  # [offset]
        tSA = tS + tA
        Sout.append(tSA)
    # print('Sout', Sout)
    return Sout


def chosenTupleId(id, n):
    intSum = 0
    for i in range(0, n):
        intSum += cpt[i][1]
        if intSum > id:
            return i
    print("error in chosenTupleId")
    return -1


def offsetID(chosen):
    intSum = 0
    if chosen == 0:
        return 0
    for i in range(0, chosen):
        intSum += cpt[i][1]
    return intSum


def lookup(tuple):
    return random.randint(1,6)


def lookupTuple(tuple):
    return "Z"


Y = ["a", "b", "c", "d"]
x = len(Y)
H = 0
sampleIndex(Y, H, x)
